Add systemd support for dhtnode and dhtdump

nemanja-boric-sociomantic opened the pull request:
This commit removes upstart scripts and changes the package to use
systemd init system instead of upstart. Note that the packages
installed from this point will not include any support for upstart.

Fixes #3

nemanja-boric-sociomantic commented on README.rst:
```diff
@@ -27,30 +27,42 @@ Many machines run multiple instances of the dhtnode. There should be a directory
+To do so, create a file ``/etc/systemd/system/dht@1.service.d/override.conf``
+with the following content:
+
+.. code::
```
seems like my rst-foo is not up to date, I'll look into this file's format.

leandro-lucarella-sociomantic replied:
```rst
.. code:: ini
  [Unit]
  ...
```

?

nemanja-boric-sociomantic replied:
Ah, missing newline.

leandro-lucarella-sociomantic commented:
Interesting: https://access.redhat.com/blogs/766093/posts/3169121

We could try the `PrivateTmp` feature.

mathias-lang-sociomantic commented:
Needs a rebase I assume :P

mathias-lang-sociomantic commented:
> Note that the packages installed from this point will not include any support for upstart.

So there is no way to have both available, or it was just done for convenience ?

mathias-lang-sociomantic commented on pkg/dhtnode-common.pkg:
```diff
@@ -5,18 +5,14 @@ OPTS.update(
     [
         "README.rst=/usr/share/doc/" + VAR.shortname + "/README.rst",
+        "deploy/systemd/dht@.service=/lib/systemd/system/dht@.service",
+        "deploy/systemd/dhtdump@.service=/lib/systemd/system/dhtdump@.service",
```
Did you look into `--deb-systemd FILEPATH        (deb only) Add FILEPATH as a systemd script` ?

mathias-lang-sociomantic replied:
Ah [I see why](https://github.com/jordansissel/fpm/issues/1363).
Would be nice to add a comment... And fix the bug ? Because that'll allow to keep both upstart and systemd script IIUC.

nemanja-boric-sociomantic replied:
:-)

nemanja-boric-sociomantic commented:
> So there is no way to have both available, or it was just done for convenience ?

I'm not sure (read: I don't remember). There still should be both directories present (and at least `/etc/init` on the systemd enabled systems), so maybe that could work.

gavin-norman-sociomantic commented:
Is there anything needs to be done to this PR? I'm thinking about getting v1.10.3 ready for deployment in the next week or so.

nemanja-boric-sociomantic commented:
I don't think this should be done in the patch release anymore though, as we solved the upstart problems.

gavin-norman-sociomantic commented:
Actually, my comment on #31 was meaningless (I just deleted it even). This isn't a patch release in the normal sense, as v1.10.x hasn't been deployed at all. So v1.10.3 is just extra features that we didn't think of when we tagged v1.10.2.

nemanja-boric-sociomantic commented:
Ah, ok. I could try to make it work for both upstart and systemd, so that we don't need systemd to run v1.10.x of the DHT node.

gavin-norman-sociomantic commented:
Oh. I thought infrastructure said it wasn't possible to have both installed?

nemanja-boric-sociomantic commented:
No, it isn't possible to do that, but it's possible for the package to support both init systems.

gavin-norman-sociomantic commented:
Gotcha. I don't mind just going for systemd, with a gradual roll-out across DCs.

nemanja-boric-sociomantic commented:
Ok, then I'll get this one sorted today.

gavin-norman-sociomantic commented:
After some discussion, it was decided that it would be better to support both init systems.

nemanja-boric-sociomantic commented:
Updated to add back upstart support and rebased.

nemanja-boric-sociomantic commented:
travis slow. SAD!

mathias-lang-sociomantic commented on deploy/systemd/dht@.service:
```diff
@@ -0,0 +1,48 @@
+ExecReload=/bin/kill -HUP $MAINPID
+
+# Not sure about this one
+#ExecReload=/bin/kill -HUP $MAINPID
```
Redundant

mathias-lang-sociomantic commented on deploy/systemd/dht@.service:
```diff
@@ -0,0 +1,48 @@
+# CPUAffinity=
+
+# Also could be interesting
+# Environment=
```
Should we keep those comments in the distributed file ?

nemanja-boric-sociomantic replied:
No, I forgot to remove them.

mathias-lang-sociomantic commented on deploy/systemd/dht@.service:
```diff
@@ -0,0 +1,48 @@
+SendSIGKILL=no
+Restart=on-failure
+User=dhtnode
+Group=core
```
Note that your package is not creating the user / group.

nemanja-boric-sociomantic replied:
How's that? 

```
if [ "$1" = "configure" ]; then
    addgroup --system core
    adduser --system --no-create-home dhtnode

```

nemanja-boric-sociomantic commented:
Updated with removal of the comments from the files.

gavin-norman-sociomantic commented:
The commit message is out-dated:
>This commit removes upstart scripts

nemanja-boric-sociomantic commented:
Updated. I'm just in the process of retesting on both systems.

nemanja-boric-sociomantic commented:
Testing showed two unaddressed points:

- There's nothing that does `systemctl enable dht@{1..4}`. This should at least be documented.
- The binary used is `/usr/sbin/dhtnode`, which is not provided by any package. I think we should probably just default to `/usr/sbin/dhtnode-1` and then use override files for the D2 testing again. Once we again decide we should go with the D2 only, we can fix the package. I don't think package can provide override files, since they are generated per instance. Sure, it could scan fs at the install time in the post-install script, but that could change. This also should be documented.

nemanja-boric-sociomantic commented:
(testing on the trusty didn't show any regressions, btw)

gavin-norman-sociomantic commented:
>The binary used is /usr/sbin/dhtnode, which is not provided by any package.

Oh, can't it run the symlinks in `/srv/dhtnode/dhtnode-*`?

nemanja-boric-sociomantic commented:
No, it can't. I'll paste now section from the previous PR.

nemanja-boric-sociomantic commented:
`ExecStart=` must contain fixed path because the executable name must not be dynamic. So, this means that we must start `/usr/sbin/dhtnode` and not `/srv/dhtnode/dhtnode-%i/dhtnode`. This causes us to create symlink in `/usr/sbin/` and not in the instance directory.

If we want to run the different binary for an instance, we can create

```
/etc/systemd/system/dht@1.service.d/override.conf
```

with the following content:

```
[Unit]
Description=D2 instance of my cool project
[Service]
# Need to clear the previous one
ExecPath=
ExecPath=/srv/dhtnode/dhtnode-1/dhtnode-d2 -c /srv/dhtnode/dhtnode-1/etc/config.ini
```

(or run `sudo systemctl edit dht@1` which will bring the editor for us).

nemanja-boric-sociomantic commented:
Ah, well now I see what I wrote previously - the idea is to create symlink somewhere else (`/usr/sbin/dhtnode`), but we could do it in `/srv/dhtnode`).

gavin-norman-sociomantic commented:
I see.

That makes switching back and forth between systemd / upstart much trickier. They'll both require different deployment configuration, not just different init files.

nemanja-boric-sociomantic commented:
I don't think it's that much of a problem, though, as they don't prohibit you having both configurations ready. Hopefully, we won't the switch more than once :-)

gavin-norman-sociomantic commented:
Yeah. It makes me wonder whether it's really worth supporting both init system at once. But whatever.

mathias-lang-sociomantic commented:
> Yeah. It makes me wonder whether it's really worth supporting both init systems at once. But whatever.

Other people will absolutely need that (because some server are shared between applications), so if you pioneer it, we'd be very grateful that you sort those issues out :)
As for the symlink, I think I mentioned `update-alternatives` when we first discussed about d1/d2 packages. So I'm going to mention it again.

nemanja-boric-sociomantic commented:
This is ready to go as far as I'm concerned.

gavin-norman-sociomantic commented:
Merginating.

gavin-norman-sociomantic commented:
Oh wait, should be based on v1.10.x. (I think we went back and forth on this.)

nemanja-boric-sociomantic commented:
Sure.

nemanja-boric-sociomantic commented:
I'll rebase.

leandro-lucarella-sociomantic commented on README.rst:
```diff
@@ -39,19 +40,59 @@ dhtnode and the former starts all instances configured on the server. Configured
+they can be used to start any number of enabled instances on the machine. Enabling
+instances is done via ``sudo systemctl enable dht@N`` where ``N`` is the instance
+number. To enable first 8 instances, one can use its shell and do
+``sudo systemctl enable dht@{1..8}``.
```
Enabling instances will make them start automatically on every reboot?

nemanja-boric-sociomantic replied:
Yes. `enable` will hook up the service into the given (multi-user) target so it will start it automatically as soon as multiuser target is up and the dependencies are up. It will not start the services, for that you need `systemctl start dht@{1..8}`
